### ISSUE TITLE
Returning "error" in the notify callback

### ISSUE THREAD

**Issue by judgej** [thephpleague/omnipay-sagepay#16]
SagePay will call up the notify URL with the results of the transaction. It includes a signature of a combination of the results, and data that was sent with the original payment request. This way if the notify results have been compromised, then it can be detected.

If the callback data has been compromised, then the application is supposed to return INVALID, along with some optional error details.

Now, the callback is handled in the application controller by calling `$gateway->completePurchase([...])->send();`. This "sends" the data we have saved for the transaction to OmniPay, which then compares it to what SagePay has just posted. If the signature does not match, then an exception is raised. This is the root of the problem.

If the signature *does* match, then it can be trusted. The `Message\ServerCompleteAuthorizeResponse` object will be returned. This object contains the methods for returning an OK, ERROR or INVALID back to SagePay, before exiting.

Now, if an exception is raised in the original send() due to a signature mismatch, then `ServerCompleteAuthorizeResponse` is never returned to the controller, and so `$response->invalid()` can never be called, and SagePay will never get to know the signature was compromised.

Instead of throwing an exception, if the signature does not match, the success status of the action should be set to false and the normal `ServerCompleteAuthorizeResponse` returned.

Similarly, right at the start of the notify controller handler, the transaction details is fetched from the database using the unique `VPSTxId` or `VendorTxCode` POST parameter. If this transaction does not exist at all, or is in a state indicating a notification for it has already been received, or it failed validation at the original submission stage, then the same `INVALID` result needs to be returned (or maybe `ERROR` at a stretch). Access to `ServerCompleteAuthorizeResponse::invalid()` is again needed, but how do we get access to this class?

*tl;dr version: if the security signature shows that the notification message has been compromised or changed en-route, the application may still want to return a response to SagePay to indicate that this has happened. With the exception currently being thrown, the response object is not returned to the application to allow this to happen.*

**Comment by judgej** [comment]
Summary of the problem, in case it is not clear:

In the callback `$gateway->completePurchase([...])->send();` must not raise an exception when the signature does not match. The callback needs to be able to return the correct response to the caller, and that is done by the `ServerCompleteAuthorizeResponse` message. When `send()` throws an exception, the callback controller never gets returned a `ServerCompleteAuthorizeResponse` to send that HTTP response to the caller. My fudgy workaround above gets around this, but it should not need to.

**Comment by simnom** [comment]
Do you think this is as simple as replacing (in Message/ServerCompleteAuthorizeRequest):

```
if (strtolower($this->httpRequest->request->get('VPSSignature')) !== $signature) {
  throw new InvalidResponseException;
}
```

with:

```
if (strtolower($this->httpRequest->request->get('VPSSignature')) !== $signature) {
  $this->httpRequest->request->set('Status', 'INVALID');
}
```

This would provide the response object back allowing for `isSuccessful`, `isError` and `isInvalid` to be handled.

**Comment by judgej** [comment]
Arguably it could be as simple as that, but there are some subtleties.

The Status is what is sent to your merchant site, so if you overwrite that, you are potentially losing some debug information you may want to log. The driver requires the merchant site callback to return either `$response->invalid()` or `$response->confirm()`, so that state does need to be available to the callback to make that decision.

So if you set the Status to Invalid, then `$response->isSuccessful()` will be false, which is what you want. However, if the security token fails its check, I don't want to treat this as a standard not-successful transaction. When this happens, we *cannot trust* the data at all. All we should do it log it, send back a response to say "urgh" (i.e. INVALID using `$response->invalid()`) but NOT use that information to update any transactions we have pending.

So I would propose storing the token validation check a separate property and perhaps making it available through `$response->isValid()`. The `$response->isSuccessful()` could also check `isValid()` as a pre-check on the remaining data.

What do you think?

**Comment by judgej** [comment]
So your logic in the notification callback would be roughly:

    // Note that the transaction_reference contains three pieces of information as JSON
    // as constructed by OmniPay after requesting the authorisation right at the start.
    $response = $gateway->completePurchase([
        'transactionId' => $transaction_id,
        'transactionReference' => $transaction_reference,
    ])->send();

    if (!$response->isValid()) {
        // Log this stuff
        $response->invalid($return_url, 'Security token invalid');
    }

    if (!$response->isSuccessful()) {
        // Mark the transaction ($transaction_id) as done and authorised.
    } else {
        // Mark the transaction as not authorised.
    }

    // Always return confirm if we are happy with the message, regardless of
    // what the authorisation status was.
    $response->confirm($return_url);

In addition, right at the start we need to get hold of the transaction_id, which is the VendorTxCode POSTed to the notification. If that cannot be found, or it does not match a pending transaction on store, then you need to return an error. This is before you have a response to sent it through. This is what I do if I cannot identify a pending transaction:

    $request = $gateway->completePurchase([]);
    $response = new \Omnipay\SagePay\Message\ServerCompleteAuthorizeResponse($request, []);
    $response->error($return_url, 'No pending transaction'); // exits here

That feels kludgy. I need to create the response in three places, manually twice:

1. Manually if a transaction cannot be identified to return an ERROR.
2. Through completePurchase()->send() to gather and check the data sent.
3. Manually if step two raises an exception, in order to return an INVALID.

Ideally it should only need to be done once, through `send()`. The fix proposed should get rid of number 3.

**Comment by simnom** [comment]
Certainly agree that adding the token check to a separate property would be the best bet, that allows the response object to remain clean and untouched, which, as you say, could help with debugging.

Was looking at the `ServerCompleteAuthorizeRequest::getData()` method wondering if passing the transaction ID was a hard and fast requirement at that point.  But, of course, that's irrelevant as you'd still need that ID to look up the transaction reference that's been stored when the transaction was registered.

So based on your loose example, getting rid of step 3 is certainly attainable with a property check.  

My thoughts on step 1 for the missing transaction ID revolve around using a more "traditional" post validation and just sending something back without involving Omnipay.

Some something like:

```
$post_array = $this->form_validation->import_post($this->input->post(), 'callback');
if (!isset($post_array['txId'])) {
    //send back a manual response via a view or similar
}
```

Again, this is hardly ideal.

**Comment by judgej** [comment]
I'll see if I can put something together locally as my ideal way of handling the callback, and see what comes out of that.

**Comment by simnom** [comment]
OK, had a little play around with this on the side.  In short what I've done is, kept the signature check in `ServerCompleteAuthorizeRequest::getData()`, but add result to a protected property.

Created `ServerCompleteAuthorizeResponse::isValid` to check signature and return boolean. I'm debating whether this should `\Message\Response` in line with the `isSuccessful()` and `isError()` methods.

I've forked and created a branch if you want to have a peep - https://github.com/ib3ltd/omnipay-sagepay/tree/protocol3-dev

Then in my callback method I'm just getting the order from the post information and then sending the `completePurchase` request, checking `isValid`, `isSuccessful` and so on.  Still can't think of a way to get passed sending the `error()` response without manually creating the response object.  

Quick Gist here - https://gist.github.com/simnom/712a13c4f6d8edc5a987.

I'm just running a few end to end tests on it at the moment but it looks promising. Appreciate time is precious but wondered if you had any thoughts on this approach if you have a moment.

Thanks

**Comment by judgej** [comment]
That branch looks, on the whole, good to me:

https://github.com/ib3ltd/omnipay-sagepay/commit/860b5bf66ce148d0da83b068cb36b3fceedf3e95

However, it is a backward compatibility (BC) break. The callback functions would need to be changed, albeit *simplified*. I also suspect there may be many callbacks not even handling it correctly now *because* of the exception, but a BC break nonetheless.

So how do we handle this?

* Leave the functionality as it is, but update the documentation with the workaround and a clear explanation as to why it is needed.
* Raise the major version to "3" and release it as a new BC-break version.
* Make the feature "switchable", so the new functionality (i.e. not raising an exception) can be explicitly turned on.
* Any other approaches?

None of these are ideal. The major version is linked to the OmniPay core version, so that is a non-started. The documentation needs a good update, with examples, anyway. The switchable feature feels like a hack, but is probably a step in the right direction.

My vote would be for making it swiitchable. All the functionality in that branch would be there, but in `getData()` if the "doRaiseExceptionOrWhatever" property is true (the default) then an exception is immediately raised if the signing does not match. We can set that property to false, to change the behaviour to something closer to what I would hope omnipay-sagepay 3.0 would do :-).

How does that sound?

**Comment by judgej** [comment]
Fixed in PR #69 - use the `acceptNotification` handler to avoid exceptions when a response still has to be sent back.
